### DISCUSSION TITLE
Feedback on the O3DE Release 22.10.0 (Feedback due by 10/31)

### DISCUSSION THREAD

**Discussion by tonybalandiuk** [o3de/sig-release#101] [General]
This discussion will serve to collect feedback on the O3DE 22.10.0 release.
To help make your comment clear, please start out your comment with one of the following phrases:
"Something that went well..."
"Something that didn't go well..."
"Something that we should change..."

**Comment by tonybalandiuk** [comment]
From @Ulrick28  Something that we should change: "Let's re-examine the build rules. I think it's a good idea for the next stabilization branch we do a nightly build, regardless of changes. Noone is going to care that the installer is still the same from night to night as all they see to download is the most recent installer. This nightly build would catch any dependency changes that don't trigger a build. "

**Reply by tonybalandiuk** [reply]
more background on this. Today stabilization is only build when there are commits. regarding dependencies: Updating the 3p-package-source repo won't directly kick off a branch build. However adding an upgraded or new 3rd party dependency would required a config file change in o3de and this would kick off a build.

**Reply by tonybalandiuk** [reply]
The risk here is we get to the end of stabilization, and there's a 3P change that does not trigger a build. The process to update a 3P library, though, is to update the config file. There could be things we are not thinking about that could break a build. So this proposal is really a safety net. @amzn-changml is there a large financial impact to doing 1 extra build a day during the stabilization period?

**Reply by tonybalandiuk** [reply]
If the cost is low @amzn-changml then sig-release is in support of adding a nightly build for the stabilization branch.

**Reply by amzn-changml** [reply]
Yes, the cost is minimal if we do it nightly. 

Note that there are changes to the O3DE repo whenever a 3P package change is made, as the package reference to a new version is updated in this path: https://github.com/o3de/o3de/tree/development/cmake/3rdParty/Platform. 

This will trigger a nightly build, provided the change was made in the development or stabilization branch.

**Reply by vincent6767** [reply]
Thanks @amzn-changml ! This is solid action items for the next release @tonybalandiuk

**Reply by tonybalandiuk** [reply]
[Done] We have added this to the task here https://github.com/o3de/o3de/issues/13208 that will be cloned for the next release.

**Comment by lmbr-pip** [comment]
Something that we should change is that stabilization branch also needs AR metrics and some form of automated alarm when success rate % falls below x. AR runs failed a lot on stabilization but it was unclear what the expectation should be and who should be keeping an eye on the stability of the build.

**Reply by tonybalandiuk** [reply]
There's 2 things here
1) AR metrics, alarm, etc. Sig-release recommends having at least the same metrics as development because contributors will be used to seeing these and understanding them. We should also consider additional metrics that may specifically benefit stabilization efforts. @amzn-changml Can you advise what you think is a reasonable next step (what we can easily do for the next release?) vs incremental improvements?
2) who should keep an eye on the build stability? discussed in sig-release meeting. We have to think about this more. Option A: to alert "Everyone" runs the risk of nobody actually doing anything and people ignoring repeated alarms. Option  B: A specific sig is alerted, and they triage the issue and assign it to the responsible sig. Need to think about this more -feedback welcomed on options or please propose option C.

**Reply by vincent6767** [reply]
@tonybalandiuk will post this on the Discord channel to get everyone feedback on (2)

**Reply by brianherrera** [reply]
sig-build is looking into setting up a mechanism to assign issues created from build failures to specific sigs. This can help with (2). 

This was originally requested with the periodic builds, but can also work with stabilization branches.

**Reply by amzn-changml** [reply]
Yes, as @brianherrera mentioned, there are ongoing efforts to improve/augment the alert/alarms on build stability, specifically for targeting codeowners of each SIG. This is targeted to complete by next release.

We do need additional metrics around the testing of the stabilization installer, though. While we do manual and automated testing on all our targets (profile, debug, release monolithic), there are currently minimal indicators of performance and stability from a direct installation currently. Some of this can be developed in the development branch, but a direct end to end workflow for stabilization would be desirable.

**Reply by vincent6767** [reply]
> We do need additional metrics around the testing of the stabilization installer, though.

Who can help with this? @amzn-changml 
CC: @tonybalandiuk

**Reply by amzn-changml** [reply]
We'll need the involvement of SIG-Build and SIG-Testing for the task of producing the tests. We do need a owning group to establish performance and stability thresholds, though some of this work may be repurposed from the existing testing framework if some thresholds already exist. This may be cross cutting with the TSC as well

**Reply by vincent6767** [reply]
> We'll need the involvement of SIG-Build and SIG-Testing for the task of producing the tests.

Anything that we (SIG Release) can help to make happen so we can have it before the next release? @amzn-changml

**Reply by tonybalandiuk** [reply]
@vincent6767 please create a sig-build issue for "We do need additional metrics around the testing of the stabilization installer"  and link it from here.

**Reply by vincent6767** [reply]
@amzn-changml I have created a GHI to track this effort: https://github.com/o3de/o3de/issues/13307.

**Reply by tonybalandiuk** [reply]
[Done] action taken, issue has been created.

**Comment by lmbr-pip** [comment]
Something that we should change is there should be clear documentation and automation for issues seeking exception for release. Items tagged in a certain way should kick off workflow that auto attaches checklist, documentation links and next steps to take.

Folks had a lot of uncertainty about how to follow the process and it meant a lot of back and forth messaging to find out what the steps where, to get approvals etc.

**Reply by tonybalandiuk** [reply]
discussed in sig-release meeting 11/8
There are 2 concerns being raised
1) clear documentation - the current process is going to be documented here https://github.com/o3de/sig-release/issues/43
2) automation - We agree this is a good next step in improving the exception process. We will create a new issue for this. In creating the issue, we need to consider the different phases that stabilization goes through (where features require approval and bugs don't , then at a later phase even bugs may require approval)

**Reply by tonybalandiuk** [reply]
@AMZN-daimini to post a discussion to address (2) and link it here

**Reply by AMZN-daimini** [reply]
Discussion is up https://github.com/o3de/sig-release/discussions/106
Topic is a little broader, but the first two points (Discovery and Tracking) both should cover this concern.

**Reply by tonybalandiuk** [reply]
[Done] action taken, discussion has been created.

**Comment by lmbr-pip** [comment]
Something that went well, I thought @tonybalandiuk and @vincent6767 did an excellent job of communicating as release managers. Vincent esp had great bias-for-action and proactively pinged SIG chairs for things needed.

**Reply by vincent6767** [reply]
Thanks @lmbr-pip !

**Reply by tonybalandiuk** [reply]
[Done] Thanks! @lmbr-pip

**Comment by lmbr-pip** [comment]
Something to change, we should have release expectations for SIG(s) written down and shared that include expectations around roadmap issues, feature matrix, exception approvals etc. https://github.com/o3de/community/issues/161 and how it relates to release process.

We seem to be working off tribal knowledge and as SIG-chair(s) are volunteer positions, we should be very clear about expectations so folks can plan their time/contribution effort accordingly.

**Reply by vincent6767** [reply]
@lmbr-pip By release expectations do you mean what each SIG plan to release on a particular version or what SIG's responsibilities are during the release?

**Reply by tonybalandiuk** [reply]
@tonybalandiuk to create issue for this and work on it after posting V1 of the current release process

**Reply by tonybalandiuk** [reply]
[Done] issue created to capture this request https://github.com/o3de/sig-release/issues/112

**Comment by hultonha** [comment]
Something that we should change is making the exception process a little more streamlined. The GitHub board helped to triage exceptions and find them, but chasing many sig chairs was time consuming and stressful. Setting-up a communication channel for sig chairs to monitor might be a possible solution. People do not respond to tags/@ on the issue which results in lots of manual chasing (DMs/calls etc...)

**Reply by tonybalandiuk** [reply]
@AMZN-daimini please include in your discussion about streamlining the exception process, and link to it here

**Reply by AMZN-daimini** [reply]
I have added this issue (3. Notification) to the discussion and proposed a possible solution similar to the one above. Feel free to comment on it or add an alternative solution if you have ideas!
https://github.com/o3de/sig-release/discussions/106

**Reply by tonybalandiuk** [reply]
[Done] added to #106 above

**Comment by hultonha** [comment]
Something that went well was the overall communication and updates from sig-release (especially big shout out to @tonybalandiuk for the regular communication and updates).

**Reply by tonybalandiuk** [reply]
[Done] Thanks Tom!

**Comment by tonybalandiuk** [comment]
Something we should change - address the build jobs that were red for the merge to o3de-atom-sampleviewer repo. Potentially separate this merge from the editor release

**Reply by vincent6767** [reply]
@tonybalandiuk Do you know who owns the O3De-atom-sample-viewer repo? I agree we need to separate out the release with O3DE. Few questions
1. How often do we need to release the ASR? If it's not, we can do it together with O3DE. They're independent of each other to make it simpler.
2. We should ask sig-graphics-audio why the ASR's build is red though. Have we followed up on this?

**Reply by tonybalandiuk** [reply]
@vincent6767 to discuss with sig-graphics

**Reply by tonybalandiuk** [reply]
@vincent6767  do you have an update here?

**Reply by vincent6767** [reply]
Sorry. Not yet. Will follow up this week.

**Reply by vincent6767** [reply]
Hi @galibzon and @thefranke 
On the last 22.10.0 release, the SIG release was helping with the ASV release. Since this is a standalone product and not critical to the O3DE release in general, we would like to recommend SIG Graphics Audio as the project owner to also handle the release. Thoughts?

**Reply by galibzon** [reply]
@vincent6767 , yes sig-graphics-audio can and should manage ASV releases.

**Reply by vincent6767** [reply]
Thanks @galibzon ! SIG Release won't manage ASV starting from the next release then. @tonybalandiuk Is there some documentation that needs to be updated with this decision?

**Reply by tonybalandiuk** [reply]
@vincent6767  I think this discussion is a good document the decision. I will also update this in the documentation that I'm working on right now to reflect this. Thanks!

**Reply by tonybalandiuk** [reply]
[DONE] noted as a todo under release doc task https://github.com/o3de/sig-release/issues/45

**Comment by Ulrick28** [comment]
Something that should change. Atom Sample Viewer should be it's own release, unrelated to O3DE releases due to size of project and issues with build pipeline customizations. This product is also not owned by the release sig and instead is owned by sig-graphics-audio

**Reply by tonybalandiuk** [reply]
@vincent6767 this is related to the above.

**Reply by tonybalandiuk** [reply]
[Done] as noted above @Ulrick28  - SIG-Release will no longer manage releases of ASV

**Comment by brianherrera** [comment]
Something that we should change is to run the AR for the merge PRs the day before the scheduled launch. This will help avoid issues with unreliable tests or any other failures that are encountered during the AR and delay the launch. Also a clean build is typically required due to the number of changes getting merged in which takes around 3 hours to complete. 

Something that we should change regarding ASV, is to have the same level of review for branch health if we keep it in sync with the O3DE launch. For example, the main branch for ASV was already red prior to the release merge so it was difficult to determine if the failures we encountered should be addressed or ignored. I do agree with the comments that its launch should be separated.

**Reply by vincent6767** [reply]
Agree to run the AR before and makes sure the build is green before the launch day. 

> Something that we should change regarding ASV, is to have the same level of review for branch health if we keep it in sync with the O3DE launch. For example, the main branch for ASV was already red prior to the release merge so it was difficult to determine if the failures we encountered should be addressed or ignored. I do agree with the comments that its launch should be separated.

I think this needs to be followed up with SIG Graphics Audio as it belongs to each SIG to maintain the branch health https://github.com/o3de/sig-release/discussions/101#discussioncomment-4132812

**Reply by tonybalandiuk** [reply]
(1) regarding AR being run in advance, we have added that to the description of this release task, https://github.com/o3de/o3de/issues/11098  This task will be copied forward for the next release.
(2) @vincent6767  if we keep ASV in the release process, perhaps we need to enforce the health of the repo. Include in your discussion with sig-graphics

**Reply by tonybalandiuk** [reply]
@galibzon  FYI there were concerns about the health of ASV repo this last release 22.10.0   Please read the discussion here since sig-graphics-audio will own releasing ASV. Thanks.

**Reply by tonybalandiuk** [reply]
[Done] (1) added to release task (2) @galibzon notified for future releases.

**Comment by tonybalandiuk** [comment]
Something that didn't go well...We had some late work come into stabilization that ended up stranded there -meaning the code only lives in stabilization and no other branch
See thread here https://discord.com/channels/805939474655346758/816043601754325023/1034970558275780631

**Reply by vincent6767** [reply]
@tonybalandiuk Have we ever known why it happens? Isn't a review needed before they can merge into the stabilization branch?

**Reply by tonybalandiuk** [reply]
@vincent6767  it happened because our process currently relies on people knowing and following the manual process. This one slipped through. It would be better if the stabilization process was driven by a process that wasn't manual

**Reply by vincent6767** [reply]
> This one slipped through. It would be better if the stabilization process was driven by a process that wasn't manual

To confirm. Do you mean there are reviewers who approve it?

**Reply by tonybalandiuk** [reply]
I am assuming the people involved in this PR (creation/merge) were not aware of the stabilization process. I don't think they were aware that only certain code was being allowed in, and that every PR needed multiple approvals.

**Reply by tonybalandiuk** [reply]
[Done]  We have a separate discussion about adding automation around the merge stabilization process.

**Comment by vincent6767** [comment]
Something that went well
1. On release day went quite well even though there are issues with the AR and ASV. But all of those are solvable quite easily. Thanks to the team!

Something that didn't go well
1. There is a surprise in the marketing committee on the O3DE 22.10.0. release date. This results in the blog updates done after the release.
3. Features grid is not updated on the release day. It was done a day after. This is mainly due to each late feature grid update. 

Something that we should change.
1. The way we do release notes right now. I propose the release notes are updated regularly, e.g. monthly instead of at the end of the release as people tend to miss the context of the work done months before.

**Reply by tonybalandiuk** [reply]
@AMZN-daimini - please link the release note discussion when you have it

**Reply by vincent6767** [reply]
@tonybalandiuk One question related to marketing. Do we also own the release of the O3DE websites? I assume no, but would like to double-confirm with you.

**Reply by tonybalandiuk** [reply]
@vincent6767 No, we do not own the release of the Website. It is owned by Marketing. 
Note our charter does not mention the Website https://github.com/o3de/sig-release/blob/main/governance/SIG%20Release%20Charter.md

We include an update to o3debinaries.org during our release, but that's just for the downloads.

**Reply by tonybalandiuk** [reply]
@vincent6767 
"I propose the release notes are updated regularly"
The release notes discussion is here: https://github.com/o3de/sig-release/discussions/107

If we pursue a bot, or other ongoing way to capture release notes as features are submitted then we will essentially have a constantly-updated set of release notes. At that point we can just have a query that allows the community to easily see "what are the new features with release notes since the last release" which I think will be the most scalable way to address this.

**Reply by tonybalandiuk** [reply]
[Done] I think this suggested change is captured under the release notes discussion @vincent6767 .

**Reply by vincent6767** [reply]
Got it. Thanks!

**Comment by tonybalandiuk** [comment]
All items under this discussion are either address or accounted for in other tasks.
At this point SIG-Release considers this conversation "done".
